aesd-char-driver/main.c

```diff
@@ -150,14 +184,104 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
     return retval; 
 }
 
+
+
+//llseek implementation
+loff_t aesd_llseek(struct file *filp, loff_t offset, int pos) 
+{
+    struct aesd_dev *dev = filp->private_data;
+    loff_t new_pos;
+
+    mutex_lock(&dev->lock);
+    switch (pos) 
+    {
+        case SEEK_SET:
+            new_pos = offset;
+            break;
+        case SEEK_CUR:
+            new_pos = filp->f_pos + offset;
+            break;
+        case SEEK_END:
+            new_pos = aesd_circular_buffer_total_size(&dev->buffer);
+            break;
+        default:
+            mutex_unlock(&dev->lock);
+            return -EINVAL;
+    }
+
+    // Validate the new position
+    if (new_pos < 0 || new_pos > aesd_circular_buffer_total_size(&dev->buffer)) 
+    { 
+        mutex_unlock(&dev->lock);
+        return -EINVAL;
+    }
+
+    filp->f_pos = new_pos;
+    mutex_unlock(&dev->lock);
+    return new_pos;
+}
+
+
+// adjust file offset
+static long aesd_adjust_file_offset(struct file *filp, unsigned int write_cmd, unsigned int write_cmd_offset)
+{
+    struct aesd_dev *dev = filp->private_data;
+    unsigned int total_offset = 0;
+    unsigned int i;
+
+    if (write_cmd >= AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED ||
+        write_cmd_offset >= dev->buffer.entry[write_cmd].size) 
```

Comment:
This assumes your current output offset is 0, should take into account the output offset.  This could be related to the error in your unit test.